pkg/backend/display/progress_test.go

```diff
@@ -126,11 +131,17 @@ func TestProgressEvents(t *testing.T) {
 					t.Parallel()
 
 					t.Run("raw", func(t *testing.T) {
-						testProgressEvents(t, path, accept, true, width, height, true)
+						suffix := fmt.Sprintf(".interactive-%vx%v", width, height)
+						opts := defaultOpts()
```

Comment:
Apart from the suffix, I don't understand what difference this change makes compared to the old code. It makes sense for `testProgressEvents` to be responsible for creating `Options` taking into account the parameters that are passed in, such as `IsInteractive`. 
Now some properties are set by the caller and others by the callee, what is separation of concerns here?

Reply:
`testProgressEvents` is a wrapper around generating snapshots for a given set of options. It should only be setting the options that are required for it to do its purpose: run the test, capture the output, compare it to the snapshot.

Reply:
Moving the suffix out of `testProgressEvents` makes it agnostic to the set of parameters passed in, which is helpful for the subsequent PR (#17084) as it adds a new suffix. Before `testProgressEvents` didn't quite separate concerns well enough, as it inspected and changed its behavior (where the snapshot resides) as a side effect of the options passed in.

Reply:
I would still prefer for `Options` to be set either entirely by the caller or the callee, not splitting the responsibility. For example, having `testProgressEvents` call `defaultOptions` and then adjusting the options based on what it was asked to do seems clean.

Reply:
I must decline.

The purpose of the original PR using the struct with optional values and `some()`, `valueOrDefault()` was to give the caller total control, as you suggest here, over parameters while letting `testProgressEvents` set reasonable defaults.

You requested that we simplify, so I adopted a pattern in which `testProgressEvents` does not set defaults, and instead we use a helper function to let callers obtain reasonable defaults.

Now we are asked to diverge quite far from the Go style elsewhere in Pulumi code. Giving the caller a long list of parameters, especially when they are of the same type, is difficult to read and a frequent cause of bugs, source: https://google.github.io/styleguide/go/best-practices.html#function-argument-lists

This for example has a bug, but it would be difficult for any reader to identify it without the aid of tooling which renders inlay hints showing which parameters are which:

```go
func testProgressEvents(
	t testing.TB,
	path string,
	accept bool,
	suffix string,
	color colors.Color,
	renderOnDirty bool,
	deterministicOutput bool,
	showResourceChanges bool,
	isInteractive bool,
	width, height int,
	raw bool,
) { /* ... */}

// ...


	t.Run("plain", func(t *testing.B) {
		for i := 0; i < t.N; i++ {
			suffix := fmt.Sprintf(".interactive-%vx%v-plain", width, height)
			opts := defaultOpts()
			opts.Color = colors.Never
			opts.RenderOnDirty = false
			opts.ShowResourceChanges = true
			opts.IsInteractive = true
			opts.ShowResourceChanges = true
			testProgressEvents(
				t, 
				path, 
				false, 
				suffix, 
				colors.Raw,
				true,
				false,
				true,
				true,
				width, 
				height,
				true)
		}
	})
```

Reply:
And because I linked to the style guide and have been asked to make multiple changes, I feel obligated to preempt another ask:

https://google.github.io/styleguide/go/best-practices.html#variadic-options

This style has fallen out of fashion in the Go community as it introduces significant maintenance overhead. The preference these days is to either use a struct whose zero value is the reasonable default, or if it's clumsy to rewrite code so that a zero value is reasonable (such as a boolean that makes the most sense to write as a default "true") or if "unset" is a valid option, then to make it a pointer so that the callee can interpret the absence of a value.

Reply:
I see two possible approaches here:
1. The caller has full control - it sets the `Options` and the callee just passes it along to the downstream API without any changes. I think this is what your code example tries to demonstrate but you're not actually passing `opts` to `testProgressEvents` and instead passing elements of `Options` as individual arguments, making the function signature very complex. This is bad for the reasons you stated. I also wasn't going to suggest var-args. 

I expected to see something like this:

```
			suffix := fmt.Sprintf(".interactive-%vx%v-plain", width, height)
			opts := defaultOpts()
			opts.Color = colors.Never
			opts.RenderOnDirty = false
			opts.ShowResourceChanges = true
			opts.IsInteractive = true
			opts.ShowResourceChanges = true
			testProgressEvents(
				t, 
				path, 
				suffix,
				opts,
                                 /* not much else */ );
```

2. The caller is unaware of `Options`. It's an implementation detail that only `testProgressEvents` cares about. The call site would look like this:

```
			suffix := fmt.Sprintf(".interactive-%vx%v-plain", width, height)
			testProgressEvents(
				t, 
				path, 
				suffix,
				opts,
				true, // isInteractive
                                 ... );
```

The goal is to avoid a situation where the caller creates a 'half-baked' object, passes it to the callee, which then has to figure out what's missing there (is color part of the default? should it be?)

Reply:
The callee here is a test helper and it needs to set certain options (like stdout) to capture the output. We'd be inverting control to make the caller responsible.

The example I showed above includes the 5 options different tests set, which include the color and four booleans.